contracts/base/LockAndBatchCall.sol

```diff
@@ -0,0 +1,41 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+pragma solidity ^0.8.19;
+
+import {SafeCallback} from "./SafeCallback.sol";
+import {IPoolManager} from "@uniswap/v4-core/contracts/interfaces/IPoolManager.sol";
+import {PoolKey} from "@uniswap/v4-core/contracts/types/PoolKey.sol";
+import {CallsWithLock} from "./CallsWithLock.sol";
+
+abstract contract LockAndBatchCall is CallsWithLock, SafeCallback {
+    error CallFail(bytes reason);
+
+    function _settle(address sender, bytes memory data) internal virtual returns (bytes memory settleData);
+    function _handleAfterExecute(bytes memory callReturnData, bytes memory settleReturnData) internal virtual;
+
+    /// @param executeData The function selectors and calldata for any of the function selectors in ICallsWithLock encoded as an array of bytes.
+    function execute(bytes memory executeData, bytes memory settleData) external {
```

Comment:
just my opinion and happy to keep as-is, but I think we should type settleData with `SettleConfig`

the serialization/deserialization feels cumbersome from a devex POV